### PR TITLE
Core: Support commits with DVs

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds supports for committing operations with DVs.

This work is part of #11122.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 1]
```diff
@@ -769,6 +794,82 @@ protected void validateDataFilesExist(
     }
   }
 
+  // validates there are no concurrently added DVs for referenced data files
+  protected void validateAddedDVs(
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:798]: This simply validates there is at most DV per file. There would be a separate PR to automatically rewrite conflicting DVs during retries. The goal is to make DVs fully usable before optimizing.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 2]
```diff
@@ -245,13 +250,13 @@ private PartitionSpec spec(int specId) {
...
   protected void add(DeleteFile file) {
-    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    validateNewDeleteFile(file);
     add(new PendingDeleteFile(file));
   }
 
   /** Add a delete file to the new snapshot. */
   protected void add(DeleteFile file, long dataSequenceNumber) {
-    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    validateNewDeleteFile(file);
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: I'm not quite clear if there's a case if a manifest rewrite could cause an issue here.  Is it possible that just rewriting delete manifests as part of incremental maintenance would hit the table version check?  I don't feel like we can require delete file rewrite as part of that.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: This only affects new delete files that are added to the table. I had a test in the prototype PR that verified that automatic manifest compaction after upgrading to V3 didn't have issues. Let me check if I added it here too.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: Yep, it is [here](https://github.com/apache/iceberg/pull/11495/files#diff-410ff1b47d9a44a2fd5dbd103cad9463d82c8f4f51aa1be63b8b403123ab6e0eR1571). Let me know if that's what you meant.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: That said, there are some test failures in Spark that may be related. Let me explore.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: It was another issue.

**Review by amogh-jahagirdar — COMMENTED** [review]
Still need to review tests but the code changes themselves overall look good, just some minor comments. Thanks @aokolnychyi!

File: core/src/main/java/org/apache/iceberg/BaseRowDelta.java [hunk 3]
```diff
@@ -139,6 +139,10 @@ protected void validate(TableMetadata base, Snapshot parent) {
       if (validateNewDeleteFiles) {
         validateNoNewDeleteFiles(base, startingSnapshotId, conflictDetectionFilter, parent);
       }
+
+      if (base.formatVersion() >= 3 && addsDeleteFiles()) {
```
Comment (jbonofre) [inline core/src/main/java/org/apache/iceberg/BaseRowDelta.java:143]: Do we really need the format version here ? Why not just using `addsDeleteFiles` flag ? (just wondering).

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 2 above)
Comment (jbonofre) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:259]: Thanks @aokolnychyi for the background. It wasn't clear to me the purpose of `validateNewDeleteFile()`during `add`.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 4]
```diff
@@ -271,6 +276,26 @@ private void add(PendingDeleteFile file) {
...
+  protected void validateNewDeleteFile(DeleteFile file) {
+    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    Preconditions.checkArgument(formatVersion() >= 2, "Deletes are supported in V2 and above");
+    if (file.content() == FileContent.POSITION_DELETES) {
+      Preconditions.checkArgument(
+          formatVersion() != 2 || !ContentFileUtil.isDV(file),
+          "Must not use DVs for position deletes in V2: %s",
+          ContentFileUtil.dvDesc(file));
+      Preconditions.checkArgument(
+          formatVersion() == 2 || ContentFileUtil.isDV(file),
```
Comment (jbonofre) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:288]: Same question here, why not just checking if it's DV: we know that if it's DV it's format version 3+ right ?

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 5]
```diff
@@ -769,6 +794,82 @@ protected void validateDataFilesExist(
...
+            ManifestContent.DELETES,
+            parent);
+    List<ManifestFile> newDeleteManifests = history.first();
+    Set<Long> newSnapshotIds = history.second();
+
+    Tasks.foreach(newDeleteManifests)
+        .stopOnFailure()
+        .throwFailureWhenFinished()
+        .executeWith(workerPool())
+        .run(m -> validateAddedDVs(m, conflictDetectionFilter, newSnapshotIds, dvRefs));
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:828]: For clarity I agree with @amogh-jahagirdar (`manifest` instead of `m`). I'm fine to use `DV` instead of `DeleteVector` but `manifest` here would help 😃

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 6]
```diff
@@ -769,6 +794,82 @@ protected void validateDataFilesExist(
...
+              ContentFileUtil.dvDesc(file));
+        }
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  // builds a set of data file locations referenced by new DVs
+  private Set<String> dvRefs() {
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:858]: nit: it's a private method so name is not a big deal

I think that `dvRefs()` (or `dvReferences()`) makes more sense here as it checks if the `deleteFile` is a DV. So, having DV in the method name makes sense to me.

**Review by jbonofre — COMMENTED** [review]
I did a first quick pass. I will to another round asap. Thanks @aokolnychyi great work as usual !

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 4 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:292]: This method validates input from the user and ensures it is not possible to add a DV to a V2 table.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 2 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:260]: I added more testing around manifest rewrites is Spark.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java]: We need to have `dv` in the name. If so, I'd prefer a shorter var name to avoid splitting statements on multiple lines.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java]: It was me trying to work-around Spotless formatting for closures. I updated.

File: core/src/main/java/org/apache/iceberg/BaseRowDelta.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/BaseRowDelta.java]: I've restructured this part a bit to compute dv refs as deletes are being added. No longer applies.

File: core/src/test/java/org/apache/iceberg/TestBase.java [hunk 7]
```diff
@@ -643,6 +666,22 @@ protected DataFile newDataFile(String partitionPath) {
...
 
+  protected DeleteFile fileADeletes() {
+    return formatVersion >= 3 ? FILE_A_DV : FILE_A_DELETES;
+  }
+
+  protected DeleteFile fileBDeletes() {
+    return formatVersion >= 3 ? FILE_B_DV : FILE_B_DELETES;
+  }
+
+  protected DeleteFile newDeletes(DataFile dataFile) {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestBase.java:677]: maybe `newDeleteFile`?

File: core/src/test/java/org/apache/iceberg/TestSnapshotSummary.java [hunk 8]
```diff
@@ -78,7 +78,7 @@ public void testFileSizeSummary() {
 
   @TestTemplate
   public void testFileSizeSummaryWithDeletes() {
-    assumeThat(formatVersion).isGreaterThan(1);
+    assumeThat(formatVersion).isEqualTo(2);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestSnapshotSummary.java:81]: are you planning to update these tests eventually to run against format version 3?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java [hunk 9]
```diff
@@ -719,7 +719,7 @@ public void testRewriteLargeManifestsEvolvedUnpartitionedV1Table() throws IOExce
 
   @TestTemplate
   public void testRewriteSmallDeleteManifestsNonPartitionedTable() throws IOException {
-    assumeThat(formatVersion).isGreaterThan(1);
+    assumeThat(formatVersion).isEqualTo(2);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:722]: FYI I'm updating this test class in https://github.com/apache/iceberg/pull/11485

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 7 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestBase.java:677]: Other `newDeleteFile` methods below return V2 position deletes, so I wanted to differentiate. I am OK renaming too.

File: core/src/test/java/org/apache/iceberg/TestSnapshotSummary.java (hunk 8 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestSnapshotSummary.java:81]: Not sure. There are new tests for DVs in this class. I think this one needs V2 position deletes, we simply didn't have the validation before. V2 position deletes are not allowed in V3 tables.

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java (hunk 9 above)
Comment (aokolnychyi) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:722]: Awesome, will take a look!

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 7 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestBase.java:677]: I believe callers won't care as long as you get a correct `DeleteFile` for the correct `formatVersion`, so I was thinking that we most likely don't need to differentiate here with the naming

**Review by nastra — APPROVED** [review]

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 7 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestBase.java:677]: I have no preference, to be honest. I'll update in the next round.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 10]
```diff
@@ -268,9 +274,32 @@ private void add(PendingDeleteFile file) {
...
+      }
+    }
+  }
+
+  protected void validateNewDeleteFile(DeleteFile file) {
+    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    Preconditions.checkArgument(formatVersion() >= 2, "Deletes are supported in V2 and above");
+    if (file.content() == FileContent.POSITION_DELETES) {
+      Preconditions.checkArgument(
+          formatVersion() != 2 || !ContentFileUtil.isDV(file),
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:288]: This is a little more clear if we invert the check:

```suggestion
formatVersion() == 2 && ContentFileUtil.isDV(file),
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java [hunk 11]
```diff
@@ -956,6 +956,62 @@ public void testRewriteLargeDeleteManifestsPartitionedTable() throws IOException
     assertThat(deleteManifests).hasSizeGreaterThanOrEqualTo(2);
   }
 
+  @TestTemplate
+  public void testUpgradeToV3() throws IOException {
```
Comment (danielcweeks) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:960]: I feel a little awkward about the scope of this test.  It's testing upgrade and manifest rewrite, but there are specific aspects that are included.

I feel like there are more scenarios to cover here (e.g. writing a DV to an existing file that has positional delete after the upgrade.).  I would say either refine the name to be more explicit (e.g. `testUpgradeToV3WithExistingV2Deletes`) so that we can add more test cases for the upgrade path.

Comment (aokolnychyi) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:960]: I agree the name is too generic. I did cover adding a DV to an existing file with positional delete in a previous PR [here](https://github.com/apache/iceberg/pull/11467/files#diff-f7cf300e48afd586e158fba64c058d30d14d37b3e96739d6cca23758ee444593R633)).

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 10 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:288]: Wouldn't it fail for V3 tables, though?

File: core/src/test/java/org/apache/iceberg/TestBase.java (hunk 7 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestBase.java:677]: I went to rename and I remembered why I did it this way. I wanted a way to construct a V2 delete file independently of the format version and to not need to update the other methods that create delete files now.

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java (hunk 11 above)
Comment (aokolnychyi) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java]: There is also an extra test in `TestRowDelta`. This one was for the action that rewrites manifests, but I renamed both.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 10 above)
Comment (danielcweeks) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:288]: Yeah, you're right.  I feel like these are hard to read because of the way the checks stack on each other.

**Review by danielcweeks — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @nastra @danielcweeks @jbonofre @amogh-jahagirdar!
